runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -4788,341 +4800,300 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
-   dependencies->addPostCondition(componentClassReg, TR::RealRegister::AssignAny);
-   dependencies->addPostCondition(temp2Reg, TR::RealRegister::AssignAny);
-   dependencies->addPostCondition(temp3Reg, TR::RealRegister::AssignAny);
+   TR::LabelSymbol *slowPathLabel = generateLabelSymbol(cg);
+   cursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, inlineAllocFailLabel, cursor);
+   cursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_B, node, slowPathLabel, cursor);
 
-   generateS390LabelInstruction(cg, TR::InstOpCode::label, node, oolJumpLabel);
-   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, oolFailLabel);
+   /********************************************* Merge inline and helper results *********************************************/
+   cursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, controlFlowEndLabel, dependencies, cursor);
+   // Can not have a collected reference register with unset value when helper is called. That would cause wrong GC register map.
+   TR::Register *collectedReferenceResultReg = cg->allocateCollectedReferenceRegister();
+   cursor = generateRRInstruction(cg, TR::InstOpCode::LGR, node, collectedReferenceResultReg, resultReg, cursor);
+   iComment("Copy tmp result to final result.");
 
-   generateS390LabelInstruction(cg, TR::InstOpCode::label, node, cFlowRegionDone, dependencies);
-
-   TR::Register *targetRegisterFinal = cg->allocateCollectedReferenceRegister();
-   generateRRInstruction(cg, TR::InstOpCode::LGR, node, targetRegisterFinal, targetReg);
-
-   // Generate the OOL code before final bookkeeping.
-   TR_S390OutOfLineCodeSection *outlinedSlowPath = new (cg->trHeapMemory()) TR_S390OutOfLineCodeSection(oolFailLabel, cFlowRegionDone, cg);
+   /********************************************* OOL helper setup *********************************************/
+   TR_S390OutOfLineCodeSection *outlinedSlowPath = new (cg->trHeapMemory()) TR_S390OutOfLineCodeSection(slowPathLabel, controlFlowEndLabel, cg);
    cg->getS390OutOfLineCodeSectionList().push_front(outlinedSlowPath);
    outlinedSlowPath->swapInstructionListsWithCompilation();
-   generateS390LabelInstruction(cg, TR::InstOpCode::label, node, oolFailLabel);
+   generateS390LabelInstruction(cg, TR::InstOpCode::label, node, slowPathLabel);
 
    TR::ILOpCodes opCode = node->getOpCodeValue();
    TR::Node::recreate(node, TR::acall);
-   TR::Register *targetReg2 = TR::TreeEvaluator::performCall(node, false, cg);
+   TR::Register *slowResultReg = TR::TreeEvaluator::performCall(node, false, cg);
    TR::Node::recreate(node, opCode);
 
-   generateRRInstruction(cg, TR::InstOpCode::LGR, node, targetReg, targetReg2);
-   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionDone);
+   generateRRInstruction(cg, TR::InstOpCode::LGR, node, resultReg, slowResultReg);
```

Comment:
Asking, would it be possible to copy this to collectedResultReg ? so that we do not perform copy again in the mainline?